### PR TITLE
Create secured item preview in sitecore

### PR DESCRIPTION
**Pull request by vbernard-coveo**
First final version. Ready to be reviewed.

### CODE REVIEW

File: _posts/2017-07-12-creating-secured-item-preview-in-sitecore.md [hunk 1]
```diff
@@ -0,0 +1,140 @@
+---
+layout: post
+
+title: "Creating secured items preview in Sitecore"
```
Comment (amoreauCoveo) [inline _posts/2017-07-12-creating-secured-item-preview-in-sitecore.md]: Creating secured item previews in Sitecore

File: _posts/2017-07-12-creating-secured-item-preview-in-sitecore.md [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+author:
+  name: Vincent Bernard
+  bio: Solution Architect, Coveo for Sitecore
+  twitter: vincentbernard_
+  image: vbernard.png
+
+---
+
+Recently, I had to work on an interesting use case where my client wanted to display teasers of premium content to anonymous users in Sitecore. Once the users found what they were looking for, they were either redirect to a login page or to a subscribing page. No secrets, the strategy behind is to increase conversion rates.
```
Comment (amoreauCoveo) [inline _posts/2017-07-12-creating-secured-item-preview-in-sitecore.md]: . . . either redirected to a  . . .
. . . strategy behind it is to increase . . .

File: _posts/2017-07-12-creating-secured-item-preview-in-sitecore.md [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+After working on the project, I realized that the business needs behind their request was fairly common. People want to show partial items to anonymous users, but still want them to be relevant and easy to find. Let me introduce you the `CreateSecuredItemPreview` solution.
+
+<!-- more -->
+
+## Overview
+
+Sitecore allows authors to set permission levels on items. These items won't be available to users having insufficient privileges. Coveo for Sitecore is using Sitecore’s permission scheme, so the results are returned accordingly to the Sitecore user's permission. This solution will create limited copies of selected items and will let anyone search for them without displaying sensitive information.
```
Comment (amoreauCoveo) [inline _posts/2017-07-12-creating-secured-item-preview-in-sitecore.md]: returned according to the Sitecore . . .

File: _posts/2017-07-12-creating-secured-item-preview-in-sitecore.md [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+![Assign the template to your items](/images/2017-07-12/2.png "Assign the template to your items")
+
+#### 2. Fill the configuration file
+
+Insert the CreateSecureItemPreview.config in `YOUR_SITECORE_SITE\Website\App_Config\Include\Coveo`.
+
+There are 5 sections in the file:
+
+* `coveoPostItemProcessingPipeline`: adds the CreateSecuredItemPreviewProcessor to the pipeline;
+* `CreateSecuredItemPreviewProcessor` references the `Remove Fields` and the `Remove Fields` defined in step 1, and which suffix will be used on the copied items;
```
Comment (francoislg) [inline _posts/2017-07-12-creating-secured-item-preview-in-sitecore.md]: `the `Remove Fields` and the `Remove Fields` ?

`which suffix will be used` ?

File: _posts/2017-07-12-creating-secured-item-preview-in-sitecore.md [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+* `fieldMap`: defines the required fields for the new processor. All the fields are tagged as external;
+* `formsAuthConfiguration`: configures the HTML crawler authentication used to log into Sitecore. This needs to point to the admin login (`/sitecore/login`);
+* `event`: adds a handler to delete the duplicate items when the original item is deleted (see disclaimer below).
+
+#### 3. Install the processor
+
+There are advanced explanations on how to build the processor in the Github folder. Once built, you  need to drop the DLL in `YOUR_SITECORE_SITE\Website\bin`.The processor intercepts items tagged with either the field `Remove Field` or the `Remove Preview`, and creates a copy.
```
Comment (francoislg) [inline _posts/2017-07-12-creating-secured-item-preview-in-sitecore.md]: "you  need" => "you need" (space)

Space before "The processor"

`Github folder` => `GitHub project.`, + add a direct link to the project.

File: _posts/2017-07-12-creating-secured-item-preview-in-sitecore.md [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+
+After working on the project, I realized that the business needs behind their request was fairly common. People want to show partial items to anonymous users, but still want them to be relevant and easy to find. Let me introduce you the `CreateSecuredItemPreview` solution.
+
+<!-- more -->
+
+## Overview
+
+Sitecore allows authors to set permission levels on items. These items won't be available to users having insufficient privileges. Coveo for Sitecore is using Sitecore’s permission scheme, so the results are returned accordingly to the Sitecore user's permission. This solution will create limited copies of selected items and will let anyone search for them without displaying sensitive information.
+
+#### Use cases
```
Comment (amoreauCoveo) [inline _posts/2017-07-12-creating-secured-item-preview-in-sitecore.md]: Tu passes de `##` à `####` right away. Ça serait pas mieux faire `###` à la place?

File: _posts/2017-07-12-creating-secured-item-preview-in-sitecore.md [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+
+* Hide specific fields to anonymous users, like phone numbers or price;
+* Hide premium content to unsubscribed members;
+* Let anonymous users search for certain articles, and then propose a subscription if they find what they are looking for.
+
+#### Detail of the solution
```
Comment (amoreauCoveo) [inline _posts/2017-07-12-creating-secured-item-preview-in-sitecore.md]: Same `####` issue

File: _posts/2017-07-12-creating-secured-item-preview-in-sitecore.md [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+
+This four-part solution copy selected items and make them available to anonymous users. It includes:
```
Comment (amoreauCoveo) [inline _posts/2017-07-12-creating-secured-item-preview-in-sitecore.md]: This four-part solution copies selected items and makes them . . .

File: _posts/2017-07-12-creating-secured-item-preview-in-sitecore.md [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+Searching for any content in the PDF will return both items, but the limited one won't have any preview. 
+
+![Restricted item preview](/images/2017-07-12/5.png "Restricted item preview")
+
+
+#### Disclaimer
+
+This processor has been tested in Launch Sitecore. To ensure that it works correctly in that environment, you'll have to:
+
+* Remove the `ExcludeCoveoUserAgentProcesso`. An exception was causing the HTML Content processor to crash;
```
Comment (francoislg) [inline _posts/2017-07-12-creating-secured-item-preview-in-sitecore.md]: you dropped a letter at the end of `ExcludeCoveoUserAgentProcesso`.

 `r` ノ( º _ ºノ)

File: _posts/2017-07-12-creating-secured-item-preview-in-sitecore.md [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+#### Detail of the solution
+
+This four-part solution copy selected items and make them available to anonymous users. It includes:
+
+* A set of Sitecore fields to manage the processor;
+* A configuration file to implement the processor;
+* A C# processor to duplicate the items and manage the security;
+* An extension to restrict fields or preview of copied items.
+
+#### Code
```
Comment (amoreauCoveo) [inline _posts/2017-07-12-creating-secured-item-preview-in-sitecore.md]: `###`

File: _posts/2017-07-12-creating-secured-item-preview-in-sitecore.md [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+
+The solution is available from [our Github repository](https://github.com/coveo/samples/tree/master/coveo-for-sitecore/solutions/CreateSecuredItemPreview).
+
+## How to use it
+
+Of course, all of this is done on a Sitecore server, with Coveo for Sitecore installed and a configured Coveo Cloud organization.
+
+#### 1. Create the required Sitecore fields
```
Comment (amoreauCoveo) [inline _posts/2017-07-12-creating-secured-item-preview-in-sitecore.md]: `###`

File: _posts/2017-07-12-creating-secured-item-preview-in-sitecore.md [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+
+First, create a template with a set of fields; those will be used as parameters to determine how the item should be copied. Make sure to use those exacts titles, or to adjust files constants.
```
Comment (amoreauCoveo) [inline _posts/2017-07-12-creating-secured-item-preview-in-sitecore.md]: amazing semi-colon 10/10 A++

File: _posts/2017-07-12-creating-secured-item-preview-in-sitecore.md [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+
+* `Remove Fields`: Checkbox
+* `Fields To Remove IDs`: Treelist scoped to `/sitecore/templates/`
+* `Enable Search In Removed Fields`: Checkbox
+* `Remove Preview`: Checkbox
+* `Enable Search in Removed Preview`: Checkbox
+
+I created mine in `/sitecore/Templates/User Defined`, and the final result look like this:
```
Comment (amoreauCoveo) [inline _posts/2017-07-12-creating-secured-item-preview-in-sitecore.md]: final result looks

File: _posts/2017-07-12-creating-secured-item-preview-in-sitecore.md (hunk 4 above)
Comment (amoreauCoveo) [inline _posts/2017-07-12-creating-secured-item-preview-in-sitecore.md]: why two `Remove Fields` ?

File: _posts/2017-07-12-creating-secured-item-preview-in-sitecore.md [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+
+There are 5 sections in the file:
+
+* `coveoPostItemProcessingPipeline`: adds the CreateSecuredItemPreviewProcessor to the pipeline;
+* `CreateSecuredItemPreviewProcessor` references the `Remove Fields` and the `Remove Fields` defined in step 1, and which suffix will be used on the copied items;
+* `fieldMap`: defines the required fields for the new processor. All the fields are tagged as external;
+* `formsAuthConfiguration`: configures the HTML crawler authentication used to log into Sitecore. This needs to point to the admin login (`/sitecore/login`);
+* `event`: adds a handler to delete the duplicate items when the original item is deleted (see disclaimer below).
+
+#### 3. Install the processor
```
Comment (amoreauCoveo) [inline _posts/2017-07-12-creating-secured-item-preview-in-sitecore.md]: `###`

File: _posts/2017-07-12-creating-secured-item-preview-in-sitecore.md (hunk 5 above)
Comment (amoreauCoveo) [inline _posts/2017-07-12-creating-secured-item-preview-in-sitecore.md]: extra space between `you` and `need` in the `Once built` sentence

Last sentence makes more sense like this:
The processor intercepts items tagged either with the `Remove Field` or the `Remove Preview` field, and creates a copy.

File: _posts/2017-07-12-creating-secured-item-preview-in-sitecore.md [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+The copied item will:
+
+* have the suffix, set in the configuration file, appended to its URI;
+* have an anonymous profile security level;
+* have a field indicating the item is a copy;
+* have a field indicating all the fields to be removed.
+
+The text contained in (`SearchViewSnippet.txt`) is used to modify the search interface so copied items won't show up to logged users.
+
+#### 4. Create the extension
```
Comment (amoreauCoveo) [inline _posts/2017-07-12-creating-secured-item-preview-in-sitecore.md]: `###`

File: _posts/2017-07-12-creating-secured-item-preview-in-sitecore.md [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+* saved the item.
+
+![Checking boxes in a Sitecore item](/images/2017-07-12/6.png "Checking boxes in a Sitecore item")
+
+
+Saving the item will re-index it. On the Coveo Cloud platform, the log browser will show the extension logs.
+
+![Log browser logging](/images/2017-07-12/3.png "Log browser logging")
+
+In the content browser, check the `View all` content checkbox. Two PDFs are now part of my index; the original one, and the duplicated one.
```
Comment (amoreauCoveo) [inline _posts/2017-07-12-creating-secured-item-preview-in-sitecore.md]: Content Browser (capitalized)

Also, not a good use of semi-colon :( you should use a regular colon instead

File: _posts/2017-07-12-creating-secured-item-preview-in-sitecore.md (hunk 9 above)
Comment (amoreauCoveo) [inline _posts/2017-07-12-creating-secured-item-preview-in-sitecore.md]: ExcludeCoveoUserAgentProcessor most likely ;)

File: _posts/2017-07-12-creating-secured-item-preview-in-sitecore.md [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+#### Disclaimer
+
+This processor has been tested in Launch Sitecore. To ensure that it works correctly in that environment, you'll have to:
+
+* Remove the `ExcludeCoveoUserAgentProcesso`. An exception was causing the HTML Content processor to crash;
+* Make sure you have configured a login page on your Sitecore site to redirect the un-authenticated user correctly.
+
+When you delete an item that has copies created with the HiddenButSearchable processor:
+
+* Deleting the item in the content tree will remove the item and all variations (language, version or any item with a suffix);
```
Comment (amoreauCoveo) [inline _posts/2017-07-12-creating-secured-item-preview-in-sitecore.md]: language, version, or any item with a suffix (fallait bin que je parle d'Oxford comma :D )

File: _posts/2017-07-12-creating-secured-item-preview-in-sitecore.md [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+* Deleting a version of the item will remove a specific version of the item, that's why a custom event handler is included in the solution.
```
Comment (amoreauCoveo) [inline _posts/2017-07-12-creating-secured-item-preview-in-sitecore.md]: change comma to period (otherwise it's a comma splice)

**Pull request merged**